File: leetcode/1898-maximum-number-of-removable-characters/Solution.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    public int maximumRemovals(String s, String p, int[] removable) {
        int n = s.length();
        int lo = 0, hi = removable.length;

        while (lo < hi) {
            int mid = lo + (hi - lo + 1) / 2;
            if (isSubsequence(s, p, removable, mid)) {
                lo = mid;
            } else {
                hi = mid - 1;
            }
        }
        return lo;
    }

    private boolean isSubsequence(String s, String p, int[] removable, int k) {
        Set<Integer> removed = new HashSet<>();
        for (int i = 0; i < k; i++) removed.add(removable[i]);

        int j = 0;
        for (int i = 0; i < s.length() && j < p.length(); i++) {
            if (removed.contains(i)) continue;
            if (s.charAt(i) == p.charAt(j)) j++;
        }
        return j == p.length();
    }
}

File: leetcode/1898-maximum-number-of-removable-characters/solution.py
from typing import List


class Solution:
    def maximumRemovals(self, s: str, p: str, removable: List[int]) -> int:
        n = len(s)

        def is_subsequence(k: int) -> bool:
            removed = set(removable[:k])
            j = 0
            for i in range(n):
                if i in removed:
                    continue
                if j < len(p) and s[i] == p[j]:
                    j += 1
            return j == len(p)

        lo, hi = 0, len(removable)
        while lo < hi:
            mid = (lo + hi + 1) // 2
            if is_subsequence(mid):
                lo = mid
            else:
                hi = mid - 1
        return lo
